### PR TITLE
okteto exec to work the same as kubectl exec

### PR DESCRIPTION
**Pull request by jLopezbarb**
DEV-327

# Proposed changes

In this PR we've addressed that `okteto exec` was behaving different to `kubectl exec`. Also we didn't have any test for that command, which now we can have. 

- Use the same pre-processing as kubernetes for `kubectl exec` and `okteto exec`. We still maintain the selector in case the user has more than one dev environment and doesn´t add it to the arguments
- Add unit tests for the whole command

## How to validate

1. Run `okteto up` on any okteto configured repository 
2. Run `okteto exec name -- echo test` and check that is working
3. Run `okteto exec name -- "echo test"` and observe an error

## CLI Quality Reminders 🔧

For both authors and reviewers:

- Scrutinize for potential regressions
- Ensure key automated tests are in place
- Build the CLI and test using the validation steps
- Assess Developer Experience impact (log messages, performances, etc)
- If too broad, consider breaking into smaller PRs
- Adhere to our [code style](https://github.com/okteto/okteto/blob/master/docs/code-style.md) and [code review](https://github.com/okteto/okteto/blob/master/docs/code-review.md) guidelines

### CODE REVIEW

**Comment by andreafalzetti** [comment]
Great work 👏 Code changes LGTM. I've tested different scenarios and everything worked besides the following:

1. Using: https://github.com/andreafalzetti/hybrid-test
2. `okteto up`
3. once the up session is up & running in another terminal run: `okteto exec hybridapp -l debug -- time`
4. notice how the command "time" is not executed. I've tried with `touch file.txt` too and it seems like it's a problem of not executing the command instead of a problem of "output".
5.

**Review by andreafalzetti — COMMENTED** [review]
Great work 👏 Code changes LGTM. I've tested different scenarios and everything worked for me besides the following:

1. Using: https://github.com/andreafalzetti/hybrid-test
2. `okteto up`
3. once the up session is up & running in another terminal run: `okteto exec hybridapp -l debug -- time`
4. notice how the command "time" is not executed. I've tried with `touch file.txt` too and it seems like it's a problem of not executing the command instead of a problem of "output".

**Update**

As we noticed offline, it only seems to be related to `time`. Other commands execute corretly.

**Review by andreafalzetti — APPROVED** [review]

File: cmd/exec/mixpannel.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
+// Copyright 2024 The Okteto Authors
```
Comment (ifbyol) [inline cmd/exec/mixpannel.go:1]: Nit: Single `n` in the filename? mixpanel vs mixpannel

File: cmd/exec.go [hunk 2]
```diff
@@ -20,6 +20,7 @@ import (
 	"os"
 	"time"
 
+	"github.com/kballard/go-shellquote"
```
Comment (ifbyol) [inline cmd/exec.go:23]: I think this file is not being used anymore. Now everything goes through `exec` package, right?

File: cmd/exec/cmd.go [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+	k8sClientProvider okteto.K8sClientProvider
+	mixpannelTracker  metadataTracker
+
+	getExecutorFunc func(dev *model.Dev, podName string) (executor, error)
+	onExecFinish    []onExecFinish
+}
+
+// NewExec creates a new exec command
+func NewExec(fs afero.Fs, ioCtrl *io.Controller) *Exec {
+	k8sProvider := okteto.NewK8sClientProvider()
```
Comment (ifbyol) [inline cmd/exec/cmd.go:58]: Ideally, this should come from `main.go` too. That is something common to all commands that can be created just once

File: cmd/exec/cmd.go [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+	mixpannelTracker := &mixpannelTrack{
```
Comment (ifbyol) [inline cmd/exec/cmd.go:59]: mixpanel? (single `n` instead of double `n`)

File: cmd/exec/cmd.go [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+		ioCtrl:            ioCtrl,
+		fs:                fs,
+		k8sClientProvider: k8sProvider,
+		appRetriever:      newAppRetriever(ioCtrl, k8sProvider),
+		mixpannelTracker:  mixpannelTracker,
+		onExecFinish: []onExecFinish{
+			mixpannelTracker.Track,
+		},
+	}
+	e.getExecutorFunc = e.getExecutor
```
Comment (ifbyol) [inline cmd/exec/cmd.go:72]: Maybe I'm missing something, but I don't get why this? `getExecutor` is already a function of `e`, why do we need to set a field of the struct with that function?

File: cmd/exec/cmd.go [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+This command allows you to run tools or perform operations directly on a dev environment that is already running.
+For more information on managing pods, refer to the okteto documentation: https://www.okteto.com/docs/`,
+		Example: `# Run the 'echo this is a test' command inside the pod named 'my-pod'
+okteto exec my-pod -- echo this is a test
+
+# Get an interactive shell session inside the pod named 'my-pod'
+okteto exec my-pod`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			manifestOpts := contextCMD.ManifestOptions{Filename: execFlags.manifestPath, Namespace: execFlags.namespace, K8sContext: execFlags.k8sContext}
+			manifest, err := contextCMD.LoadManifestWithContext(ctx, manifestOpts, e.fs)
```
Comment (ifbyol) [inline cmd/exec/cmd.go:94]: Does this take into consideration the env vars for context and namespace after the refactor done by andrea?

File: cmd/exec/options.go [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+import (
+	"errors"
+	"fmt"
+
+	"github.com/okteto/okteto/cmd/utils"
+	"github.com/okteto/okteto/pkg/log/io"
+	"github.com/okteto/okteto/pkg/model"
+)
+
+type Options struct {
```
Comment (ifbyol) [inline cmd/exec/options.go:25]: Nit: If this is not being used outside this package, we might not want to export it

File: cmd/exec/cmd.go [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+	e.mixpannelTracker.SetMetadata(&analytics.TrackExecMetadata{
+		Mode:               dev.Mode,
+		FirstArgIsDev:      opts.firstArgIsDevName,
+		Success:            err == nil,
+		IsOktetoRepository: utils.IsOktetoRepo(),
+		IsInteractive:      dev.IsInteractive(),
+	})
+	for _, f := range e.onExecFinish {
+		f()
+	}
```
Comment (ifbyol) [inline cmd/exec/cmd.go:155]: This is a bit odd. I mean, it feels odd that we call directly to mixpanelTracker to set some metadata to be sent, but then, we don't use it directly to track the information. 

2 alternatives to this:
* As we only have 1 tracker or just one operation to do when the command ends, I think we can remove the `onExecFinish` field in the struct and just call to the mixpanelTracker
* functions within `onExecFinish` field receive the "metadata" to be sent and remove the dependency with `mixpanelTracker` in the `ExecCommand`

With any of the cases we can remove one dependency from `Exec` struct

**Review by ifbyol — COMMENTED** [review]
There is nothing blocker but some comments would be good to addres

File: cmd/exec/cmd.go (hunk 6 above)
Comment (jLopezbarb) [reply cmd/exec/cmd.go:95]: Yes, I just tested and its working correctly with the latest changes done

**Comment by ifbyol** [comment]
I have been testing it and I found small things to improve or thing that have changed.

1. When you don't specify a dev env where to execute the command, before, we were displaying the selector with just the dev environments in dev mode. Now, we display all the devs (in that screen, the terminal in the bottom is current CLI version and the other one is the version of this branch). If you select a dev which is not in dev mode, it fails (which is fine), but I think we should display only the services that are in dev mode

Screenshot 2024-05-24 at 13 19 39

2. When a command fails the execution, we are displaying the error message (the red one) in the same line that previous errors, Could we just add a new line to show the red message in a new line? This is not new, but if it is something easy to fix

Screenshot 2024-05-24 at 13 25 08

3. When I execute `okteto exec` without specifying a service (and there are several services) and without specifying a command, I have to select the dev env first, to then, fail with an error indicating that a command is required. Could we do the validation first instead of making the user to select a dev to then, fail?

**Review by ifbyol — COMMENTED** [review]
Do we have some e2e test for the `exec` command?

**Comment by jLopezbarb** [comment]
> Do we have some e2e test for the `exec` command?

Not right now, I'll create a simple one for this

**Comment by ifbyol** [comment]
The error message when I try to execute a command for a non-existing dev container has changed with a less meaningful error and we are missing the hint. Is that expected? 

With your changes

Screenshot 2024-05-27 at 10 55 49

Without your changes

Screenshot 2024-05-27 at 10 56 46

We should not change that unless it is strictly necessary. As it was before it is giving better DX and giving hints to the user on why it failed and how it can be solved

**Comment by ifbyol** [comment]
Something similar happens when you don't specify a dev, but there are no devs on dev mode

With these changes

Screenshot 2024-05-27 at 11 00 24

Without these changes

Screenshot 2024-05-27 at 10 59 48

Specially, this error `Failed to set dev from manifest` is not meaningful and can be misleading for users

**Pull request merged**